core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRandomDistributionEndpointSelector.java

```diff
@@ -0,0 +1,144 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ * This selector selects an {@link Endpoint} using random and the weight of the {@link Endpoint}. If there are
+ * A(weight 10), B(weight 4) and C(weight 6) {@link Endpoint}s, the chances that {@link Endpoint}s are selected
+ * are 10/20, 4/20 and 6/20, respectively. If A {@link Endpoint} is selected 10 times and B and C are not
+ * selected as much as their weight, then A is removed temporarily and the the chances that B and C are selected
+ * are 4/10 and 6/10.
+ */
+final class WeightedRandomDistributionEndpointSelector {
+
+    private final List<Entry> entries;
+    private final long totalWeight;
+    private final List<Entry> currentEntries;
+    private long currentTotalWeight;
+
+    WeightedRandomDistributionEndpointSelector(List<Endpoint> endpoints) {
+        final ImmutableList.Builder<Entry> builder = ImmutableList.builder();
+
+        long totalWeight = 0;
+        for (Endpoint endpoint : endpoints) {
+            if (endpoint.weight() <= 0) {
+                continue;
+            }
+            builder.add(new Entry(endpoint));
+            totalWeight += endpoint.weight();
+        }
+        this.totalWeight = totalWeight;
+        currentTotalWeight = totalWeight;
+        entries = builder.build();
+        currentEntries = new CopyOnWriteArrayList<>(entries);
+    }
+
+    @VisibleForTesting
+    List<Entry> entries() {
+        return entries;
+    }
+
+    @Nullable
+    Endpoint selectEndpoint() {
+        if (entries.isEmpty()) {
+            return null;
+        }
+        Endpoint selected = null;
+        for (;;) {
+            long nextLong = ThreadLocalRandom.current().nextLong(currentTotalWeight);
+            // There's a chance that currentTotalWeight is changed before looping currentEntries.
+            // However, we have counters and choosing an endpoint doesn't have to be exact so no big deal.
+            // TODO(minwoox): Use binary search when the number of endpoints is greater than N.
+            for (Entry entry : currentEntries) {
+                final int weight = entry.endpoint().weight();
+                nextLong -= weight;
+                if (nextLong < 0) {
+                    if (entry.increaseCounter()) {
+                        selected = entry.endpoint();
+                    }
+                    if (!entry.isFull()) {
+                        break;
+                    }
+
+                    // The entry is full so we should remove the entry from currentEntries.
+                    synchronized (currentEntries) {
+                        // Check again not to remove the entry where reset() is called by another thread.
+                        if (entry.isFull()) {
+                            if (currentEntries.remove(entry)) {
```

Comment:
I'm still wondering about this use of list removal. If the endpoints are fixed, then intuitively it seems that nothing should be removed. Instead, we can keep track of valid / invalid entries, possibly as a boolean on the entry itself, possibly as a map to boolean. For example, I'm wondering with a high request rate, does this not cause a lot of thrashing with constant creations of CoW list?

Reply:
Or even without high request rate, it seems like if the weights look something like `1`, `1`, `2`, this list will get recreated a ton since the entries fill up every request. Maybe this isn't intended usage of this endpoint selector but it seems fragile.

Reply:
Thinking about it, `BitSet` seems like it should work great for tracking removal of indices in a list. `clear` will be extremely fast.

Reply:
> I'm still wondering about this use of list removal.

We don't pre-calculate the weight anymore, we don't need to remove the entry from the list anymore. 😄  PTAL.